File: app/lib/models/feedback.dart
class Feedback {
  final String id;
  final String category;
  final String title;
  final String message;
  final String priority;
  final String status;
  final String? adminResponse;
  final DateTime? responseDate;
  final List<String> screenshots;
  final DeviceInfo? deviceInfo;
  final bool isPublic;
  final DateTime createdAt;
  final DateTime updatedAt;

  Feedback({
    required this.id,
    required this.category,
    required this.title,
    required this.message,
    this.priority = 'medium',
    this.status = 'pending',
    this.adminResponse,
    this.responseDate,
    this.screenshots = const [],
    this.deviceInfo,
    this.isPublic = false,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Feedback.fromJson(Map<String, dynamic> json) {
    return Feedback(
      id: json['_id'] ?? '',
      category: json['category'] ?? 'general',
      title: json['title'] ?? '',
      message: json['message'] ?? '',
      priority: json['priority'] ?? 'medium',
      status: json['status'] ?? 'pending',
      adminResponse: json['adminResponse'],
      responseDate: json['responseDate'] != null 
          ? DateTime.parse(json['responseDate'])
          : null,
      screenshots: List<String>.from(json['screenshots'] ?? []),
      deviceInfo: json['deviceInfo'] != null 
          ? DeviceInfo.fromJson(json['deviceInfo'])
          : null,
      isPublic: json['isPublic'] ?? false,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'category': category,
      'title': title,
      'message': message,
      'priority': priority,
      'status': status,
      'adminResponse': adminResponse,
      'responseDate': responseDate?.toIso8601String(),
      'screenshots': screenshots,
      'deviceInfo': deviceInfo?.toJson(),
      'isPublic': isPublic,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  String get statusDisplayText {
    switch (status) {
      case 'pending':
        return 'Pending Review';
      case 'in_progress':
        return 'In Progress';
      case 'resolved':
        return 'Resolved';
      case 'rejected':
        return 'Rejected';
      default:
        return 'Unknown';
    }
  }

  String get categoryDisplayText {
    switch (category) {
      case 'bug':
        return 'Bug Report';
      case 'feature':
        return 'Feature Request';
      case 'general':
        return 'General Feedback';
      case 'complaint':
        return 'Complaint';
      case 'suggestion':
        return 'Suggestion';
      default:
        return 'Other';
    }
  }

  String get priorityDisplayText {
    switch (priority) {
      case 'low':
        return 'Low Priority';
      case 'medium':
        return 'Medium Priority';
      case 'high':
        return 'High Priority';
      case 'critical':
        return 'Critical';
      default:
        return 'Unknown';
    }
  }

  bool get isResolved => status == 'resolved';
  bool get isPending => status == 'pending';
  bool get isInProgress => status == 'in_progress';
  bool get hasResponse => adminResponse != null && adminResponse!.isNotEmpty;
}

class DeviceInfo {
  final String? platform;
  final String? version;
  final String? model;

  DeviceInfo({
    this.platform,
    this.version,
    this.model,
  });

  factory DeviceInfo.fromJson(Map<String, dynamic> json) {
    return DeviceInfo(
      platform: json['platform'],
      version: json['version'],
      model: json['model'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'platform': platform,
      'version': version,
      'model': model,
    };
  }
}

// Predefined categories
class FeedbackCategories {
  static const List<Map<String, String>> categories = [
    {'value': 'bug', 'label': 'Bug Report'},
    {'value': 'feature', 'label': 'Feature Request'},
    {'value': 'general', 'label': 'General Feedback'},
    {'value': 'complaint', 'label': 'Complaint'},
    {'value': 'suggestion', 'label': 'Suggestion'},
  ];
}

// Predefined priorities
class FeedbackPriorities {
  static const List<Map<String, String>> priorities = [
    {'value': 'low', 'label': 'Low Priority'},
    {'value': 'medium', 'label': 'Medium Priority'},
    {'value': 'high', 'label': 'High Priority'},
    {'value': 'critical', 'label': 'Critical'},
  ];
}

File: server/models/Feedback.js
const mongoose = require('mongoose');

const feedbackSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User is required']
  },
  category: {
    type: String,
    required: [true, 'Please specify feedback category'],
    enum: ['bug', 'feature', 'general', 'complaint', 'suggestion'],
    default: 'general'
  },
  title: {
    type: String,
    required: [true, 'Please add a feedback title'],
    trim: true,
    maxlength: [100, 'Title cannot be more than 100 characters']
  },
  message: {
    type: String,
    required: [true, 'Please add feedback message'],
    trim: true,
    maxlength: [1000, 'Message cannot be more than 1000 characters']
  },
  priority: {
    type: String,
    enum: ['low', 'medium', 'high', 'critical'],
    default: 'medium'
  },
  status: {
    type: String,
    enum: ['pending', 'in_progress', 'resolved', 'rejected'],
    default: 'pending'
  },
  adminResponse: {
    type: String,
    trim: true,
    maxlength: [500, 'Admin response cannot be more than 500 characters']
  },
  responseDate: {
    type: Date
  },
  screenshots: [{
    type: String, // Base64 or URL
    maxlength: [500, 'Screenshot path too long']
  }],
  deviceInfo: {
    platform: String,
    version: String,
    model: String
  },
  isPublic: {
    type: Boolean,
    default: false // Whether feedback can be shown to other users
  }
}, {
  timestamps: true
});

// Create indexes for better performance
feedbackSchema.index({ userId: 1 });
feedbackSchema.index({ category: 1 });
feedbackSchema.index({ status: 1 });
feedbackSchema.index({ priority: 1 });
feedbackSchema.index({ createdAt: -1 });

// Static method to get feedback statistics
feedbackSchema.statics.getFeedbackStats = async function() {
  const stats = await this.aggregate([
    {
      $group: {
        _id: '$status',
        count: { $sum: 1 }
      }
    }
  ]);
  
  const categoryStats = await this.aggregate([
    {
      $group: {
        _id: '$category',
        count: { $sum: 1 }
      }
    }
  ]);
  
  return { statusStats: stats, categoryStats };
};

// Static method to get pending feedbacks
feedbackSchema.statics.getPendingFeedbacks = function(limit = 10) {
  return this.find({ status: 'pending' })
    .populate('userId', 'name email shopName')
    .sort({ createdAt: -1 })
    .limit(limit);
};

// Instance method to update status
feedbackSchema.methods.updateStatus = function(status, adminResponse) {
  this.status = status;
  if (adminResponse) {
    this.adminResponse = adminResponse;
    this.responseDate = new Date();
  }
  return this.save();
};

module.exports = mongoose.model('Feedback', feedbackSchema);
